pkg/components/contour/component.go

```diff
@@ -38,38 +42,50 @@ func init() {
 // Values provided for IngressHosts is used as value for the annotation `external-dns.alpha.kubernetes.io/hostname`
 // This annotation is added to Envoy service.
 type component struct {
-	ServiceMonitor bool `hcl:"service_monitor,optional"`
 	// IngressHosts field is added in order to make contour work with ExternalDNS component.
 	// Values provided for IngressHosts is used as value for the annotation `external-dns.alpha.kubernetes.io/hostname`.
 	// This annotation is added to Envoy Service, in order for ExternalDNS to create DNS entries.
 	// This solution is a workaround for projectcontour/contour#403
 	// More details regarding this workaround and other solutions is captured in
 	// https://github.com/kinvolk/PROJECT-Lokomotive-Kubernetes/issues/474
 	IngressHosts []string `hcl:"ingress_hosts,optional"`
-
 	// IngressHostsRaw is not accessible to the user
 	IngressHostsRaw string
-
-	NodeAffinity   []util.NodeAffinity `hcl:"node_affinity,block"`
-	Tolerations    []util.Toleration   `hcl:"toleration,block"`
-	TolerationsRaw string
+	NodeAffinity    []util.NodeAffinity `hcl:"node_affinity,block"`
+	ServiceMonitor  bool                `hcl:"service_monitor,optional"`
+	ServiceType     string              `hcl:"service_type,optional"`
+	Tolerations     []util.Toleration   `hcl:"toleration,block"`
+	TolerationsRaw  string
 }
 
 func newComponent() *component {
-	return &component{}
+	return &component{
+		ServiceType: serviceTypeLoadBalancer,
+	}
 }
 
 func (c *component) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContext) hcl.Diagnostics {
+	diagnostics := hcl.Diagnostics{}
+
 	if configBody == nil {
 		return hcl.Diagnostics{
 			components.HCLDiagConfigBodyNil,
 		}
 	}
+
 	if err := gohcl.DecodeBody(*configBody, evalContext, c); err != nil {
-		return err
+		diagnostics = append(diagnostics, err...)
```

Comment:
Makes sense, but I've copied this logic as-is from existing components. We need to fix this also in the `cluster-autoscaler` and `velero` components:

https://github.com/kinvolk/lokomotive/blob/ede5d2ad41b83a09c1e857dccabf924acba4bb7c/pkg/components/cluster-autoscaler/component.go#L232-L234

https://github.com/kinvolk/lokomotive/blob/ede5d2ad41b83a09c1e857dccabf924acba4bb7c/pkg/components/velero/velero.go#L131-L133

Reply:
Opened https://github.com/kinvolk/lokomotive/issues/413 to track fixing the other components.

Reply:
Actually `gohcl.DecodeBody` returns an `hcl.Diagnostics`, not an error. The returned value should be checked using the `HasErrors()` method, not compared to nil.

Reply:
I've fixed the error check and added a return.

Passing a malformed HCL does fail as expected, however in order to add a test for this we would have to refactor the way we do component testing, which is way out of scope for this PR. The reason is - when adding a unit test with malformed HCL, `util.GetComponentBody()` fails before we have a chance to check whether we wanted to get an error or not.

Reply:
@rata can we treat this as resolved? I want to remind that we have a broken state on master and I don't want to delay merging this any further.

Reply:
Actually the return I've added doesn't do anything because we error out somewhere in `gohcl.DecodeBody()`. My error checking code never gets executed.

Reply:
I'm completely confused now. If anybody knows the proper way to handle this - please let me know. I am currently reading through the parsing logic in the HCL library :disappointed:

Reply:
@johananl if you try passing valid HCL buth with wrong structure (e.g. non existent attribute), then I think you should be able to trigger it. We have some tests, which handles that. E.g. https://github.com/kinvolk/lokomotive/blob/master/pkg/components/external-dns/component_test.go#L94

Reply:
@johananl Sure, mark as resolved when you consider it solved :)

Reply:
Thanks @invidian, that was indeed the case. I feel like we need better guidelines (and in my case - understanding) regarding how to deal with HCL errors. This still feels rather messy to me.

Reply:
It is messy indeed. @ipochi refactioring work actually encapsulates all HCL code really nicely.